mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/query/parsing/nativequery/impl/NativeQueryParser.java

```diff
@@ -57,301 +67,300 @@
  * @author Guillaume Smet
  */
 @BuildParseTree
-public class NativeQueryParser extends BaseParser<MongoDBQueryDescriptorBuilder> {
-
-	final MongoDBQueryDescriptorBuilder builder;
+public class NativeQueryParser extends BaseParser<MongoDBQueryDescriptor> {
+
+	// TODO There are many more query types than what we support.
+	protected static List<String> RESERVED = Arrays.asList(
+			"findOne", "findAndModify", "find", "insertOne", "insertMany", "insert",
+			"remove", "deleteOne", "deleteMany", "updateOne", "updateMany", "update",
+			"replaceOne", "count", "aggregate", "distinct", "mapReduce" );
+	protected MongoDBQueryDescriptorBuilder builder = new MongoDBQueryDescriptorBuilder();
+	protected CliQueryInfo cliQueryInfo = new CliQueryInfo();
+
+	private static class CliQueryInfo {
+		String operationName;
+		String collectionName;
+		List<String> pathExpression = new ArrayList<>();
+		List<String> arguments = new ArrayList<>();
+	}
+	private Action<MongoDBQueryDescriptor> processPathExpression = context -> {
+		List<String> identifiers = cliQueryInfo.pathExpression;
+		if ( identifiers.size() <= 2 ) {
+			context.getParseErrors().add(
+					createParseError(
+							context, 2, "CLI query should match 'db.<COLLECTION>.<OPERATION>'"
+							)
+					);
+			return false;
+		}
+		if ( !"db".equals( identifiers.get( 0 ) ) ) {
+			context.getParseErrors().add(
+					createParseError(
+							context, 2, "First identifier should be 'db'"
+							)
+					);
+			return false;
+		}
+		cliQueryInfo.operationName = identifiers.get( identifiers.size() - 1 );
+		cliQueryInfo.collectionName = String.join( ".", identifiers.subList( 1, identifiers.size() - 1 ) );
+		if ( RESERVED.contains( cliQueryInfo.collectionName ) ) {
```

Comment:
Are we sure that these keywords are reserved and can't be used for a collection name? (just checking)

Reply:
My bad, I'm not sure why I put this check.